SuppressMessage works in IDE but still produces warnings in build output

**Version Used**: 
VS 16.3.2
**Steps to Reproduce**:

1. Create new c# .net core Console Project with following code
   ```c#
    using System;
  
    namespace ConsoleApp2
    {
        class Program
        {
            private string test;
    
            static void Main(string[] args)
            {
                var p = new Program() { test = "bl" };
                Console.WriteLine("Hello World!" + p.test);
            }
        }
    }

   ```
2. Enable nullable in `.csproj` (field `test` will show green squigles with non initilized non-nullable warning)
3.  Add Gloabl suppression file with
  ```c#
  [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Code Quality", "CS8618:Non-nullable field is uninitialized. Consider declaring as nullable.", Justification = "<Pending>", Scope = "member", Target = "~F:ConsoleApp2.Program.test")]
  ```
  (warning in source code will vanish)

**Expected Behavior**:

Bild shows no warnings

**Actual Behavior**:
Build shows warning  `Program.cs(7,24,7,28): warning CS8618: Non-nullable field 'test' is uninitialized. Consider declaring the field as nullable.`

--- a/src/Compilers/CSharp/Test/Semantic/Diagnostics/GetDiagnosticsTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Diagnostics/GetDiagnosticsTests.cs
@@ -7,9 +7,9 @@ using System.Linq;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Diagnostics.CSharp;
 using Microsoft.CodeAnalysis.Test.Utilities;
 using Roslyn.Test.Utilities;
-using Roslyn.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
@@ -419,6 +419,44 @@ namespace N1
 
             // Verify diagnostics for the second tree. This should have triggered the assert
             compilation.GetSemanticModel(tree2).GetDeclarationDiagnostics().Verify();
+        }
+
+        [Fact]
+        [WorkItem(39094, "https://github.com/dotnet/roslyn/issues/39094")]
+        public void TestSuppressMessageAttributeDoesNotSuppressCompilerDiagnostics()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("""", ""CS0168"", Justification = """", Scope = ""type"", Target = ""~T:C"")]
+
+class C
+{
+    void M()
+    {
+        // warning CS0168:  The variable 'x' is declared but never used.
+        int x;
+    }
+}
+";
+            // Verify unsuppressed CS0168 in 'Compilation.GetDiagnostics'
+            var compilation = CreateCompilation(source);
+            var diagnostics = compilation.GetDiagnostics();
+            var expected = Diagnostic(ErrorCode.WRN_UnreferencedVar, "x").WithArguments("x").WithLocation(11, 13);
+            diagnostics.Verify(expected);
+            Assert.False(diagnostics.Single().IsSuppressed);
+
+            // Verify 'GetEffectiveDiagnostics' does not apply SuppressMessage suppression to compiler diagnostics.
+            var effectiveDiagnostics = CompilationWithAnalyzers.GetEffectiveDiagnostics(diagnostics, compilation);
+            effectiveDiagnostics.Verify(expected);
+            Assert.False(effectiveDiagnostics.Single().IsSuppressed);
+
+            // Verify CS0168 is not suppressed for compiler diagnostics computed
+            // using CompilerDiagnosticAnalyzer
+            var analyzers = new DiagnosticAnalyzer[] { new CSharpCompilerDiagnosticAnalyzer() };
+            var analyzerDiagnostics = compilation.GetAnalyzerDiagnostics(analyzers);
+            analyzerDiagnostics.Verify(expected);
+            Assert.False(analyzerDiagnostics.Single().IsSuppressed);
         }
     }
 }

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
@@ -133,14 +133,17 @@ namespace Microsoft.CodeAnalysis.Diagnostics
         }
 
         private bool IsDiagnosticSuppressed(Diagnostic diagnostic, Func<Compilation, SyntaxTree, SemanticModel> getSemanticModel, out SuppressMessageInfo info)
-            => IsDiagnosticSuppressed(diagnostic.Id, diagnostic.Location, getSemanticModel, out info);
-
-        private bool IsDiagnosticSuppressed(string id, Location location, Func<Compilation, SyntaxTree, SemanticModel> getSemanticModel, out SuppressMessageInfo info)
         {
-            Debug.Assert(id != null);
-            Debug.Assert(location != null);
+            info = default;
 
-            info = default(SuppressMessageInfo);
+            if (isCompilerDiagnostic(diagnostic))
+            {
+                // SuppressMessage attributes do not apply to compiler diagnostics.
+                return false;
+            }
+
+            var id = diagnostic.Id;
+            var location = diagnostic.Location;
 
             if (IsDiagnosticGloballySuppressed(id, symbolOpt: null, isImmediatelyContainingSymbol: false, info: out info))
             {
@@ -194,6 +197,19 @@ namespace Microsoft.CodeAnalysis.Diagnostics
                     inImmediatelyContainingSymbol = false;
                 }
                 while (namespaceSymbol != null);
+
+                return false;
+            }
+
+            static bool isCompilerDiagnostic(Diagnostic diagnostic)
+            {
+                foreach (var customTag in diagnostic.CustomTags)
+                {
+                    if (customTag == WellKnownDiagnosticTags.Compiler)
+                    {
+                        return true;
+                    }
+                }
 
                 return false;
             }

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs
@@ -136,7 +136,7 @@ namespace Microsoft.CodeAnalysis.Diagnostics
         {
             info = default;
 
-            if (isCompilerDiagnostic(diagnostic))
+            if (diagnostic.CustomTags.Contains(WellKnownDiagnosticTags.Compiler))
             {
                 // SuppressMessage attributes do not apply to compiler diagnostics.
                 return false;
@@ -197,19 +197,6 @@ namespace Microsoft.CodeAnalysis.Diagnostics
                     inImmediatelyContainingSymbol = false;
                 }
                 while (namespaceSymbol != null);
-
-                return false;
-            }
-
-            static bool isCompilerDiagnostic(Diagnostic diagnostic)
-            {
-                foreach (var customTag in diagnostic.CustomTags)
-                {
-                    if (customTag == WellKnownDiagnosticTags.Compiler)
-                    {
-                        return true;
-                    }
-                }
 
                 return false;
             }

--- a/src/Compilers/Core/Portable/InternalUtilities/IReadOnlyListExtensions.cs
+++ b/src/Compilers/Core/Portable/InternalUtilities/IReadOnlyListExtensions.cs
@@ -1,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+
+namespace Roslyn.Utilities
+{
+    internal static class IReadOnlyListExtensions
+    {
+        public static bool Contains<T>(this IReadOnlyList<T> list, T item, IEqualityComparer<T>? comparer = null)
+        {
+            comparer ??= EqualityComparer<T>.Default;
+            for (int i = 0; i < list.Count; i++)
+            {
+                if (comparer.Equals(item, list[i]))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+}

